rrd/rrd_interface.ml

```diff
@@ -32,92 +35,375 @@ let set_sockets_dir x =
-	end
-
-	module Interdomain = struct
-		external register : uid:interdomain_uid ->
-			info:interdomain_info ->
-			protocol:plugin_protocol -> float = ""
-		external deregister : uid:interdomain_uid -> unit = ""
-		external next_reading : uid:interdomain_uid -> float = ""
-	end
-
-	external register : uid:string -> frequency:Rrd.sampling_frequency ->
-		float = ""
-	external deregister : uid:string -> unit = ""
-	external next_reading : uid:string -> float = ""
-end
+(** Version 1 or 2 of plugin protocol *)
+type plugin_protocol =
+  | V1  (** Plugin protocol 1 *)
+  | V2  (** Plugin protocol 2 *)
+[@@default V2]
+[@@deriving rpcty]
 
-module HA = struct
-	external enable_and_update :
-		statefile_latencies:Rrd.Statefile_latency.t list ->
-		heartbeat_latency:float -> xapi_latency:float -> unit = ""
-	external disable : unit -> unit = ""
-end
+(** Domain ID of VM *)
+type interdomain_uid =
+  { name           : string; (** VM domain name label *)
+    frontend_domid : int     (** Front-end domain ID number *)
+  } [@@deriving rpcty]
+
+(* Note: for types such as rrd_req, which alias
+ * user-defined types, it is usually not enough
+ * to add [@@deriving rpcty] to derive RPC types,
+ * and rpc functions must be declared in the definition.
+ * However, types such as Rrd.sampling_frequency have
+ * already been defined as RPC types in their original declarations
+ * so we are able to derive these type aliases like so *)
+
+type rrd_freq = Rrd.sampling_frequency = Five_Seconds
+[@@deriving rpcty]
+
+type statefile_latency = Rrd.Statefile_latency.t = { id: string; latency: float option }
+[@@deriving rpcty]
+
+type sflat_lst = statefile_latency list
+[@@deriving rpcty]
+
+(** Domain database sampling info *)
+type interdomain_info =
+  { frequency        : rrd_freq; (** interdomain rrd sampling frequency *)
+    shared_page_refs : int list  (** list of shared page references *)
+  } [@@deriving rpcty]
+
+type string_opt = string option
+[@@deriving rpcty]
+
+type ds_list = Data_source.t list
+[@@deriving rpcty]
+
+
+(* -- error handling -- *)
+
+(** Rrdd error type *)
+type rrd_errors =
+  | Archive_failed      of string    (** Archival failure *)
+  | Invalid_protocol    of string    (** Thrown by protocol_of_string if
+                                string does not match plugin protocol *)
+  | Rrdd_internal_error of string    (** Internal Rrdd error *)
+[@@deriving rpcty]
+
+exception Rrdd_error of rrd_errors
+
+(** Error handler *)
+module RrdErrHandler = Error.Make(struct
+    type t = rrd_errors
+    let t  = rrd_errors
+    let internal_error_of e = Some (Rrdd_internal_error (Printexc.to_string e))
+  end)
+let rrd_err = Error.{ def     = rrd_errors
+                    ; raiser  = (fun e -> raise (Rrdd_error e))
+                    ; matcher = (function
+                      | Rrdd_error e -> Some e
+                      | e            -> Some (Rrdd_internal_error (Printexc.to_string e)))
+                    }
+
```

Comment:
I think you need to register a printer as we did for the other modules, to make sure that we do not loose the content of the errors:
```ocaml
let () = (* register printer *)
  let sprintf = Printf.sprintf in
  let string_of_error e =
    Rpcmarshal.marshal rrd_err.Rpc.Types.ty e |> Rpc.to_string in
  let printer = function
    | Gpumon_error e ->
        Some (sprintf "Rrd_interface.Rrdd_error(%s)" (string_of_error e))
    | _ -> None in
Printexc.register_printer printer
```